### ISSUE TITLE
1.0.39 clock_gettime problem on OS X 10.11 with Xcode 8.x

### ISSUE THREAD

**Issue by ilovezfs** [Tarsnap/tarsnap#286]
```
==> brew test tarsnap --verbose
==> FAILED
Testing tarsnap
/usr/bin/sandbox-exec -f /tmp/homebrew20170821-9565-xa46nn.sb /System/Library/Frameworks/Ruby.framework/Versions/2.0/usr/bin/ruby -W0 -I /usr/local/Homebrew/Library/Homebrew -- /usr/local/Homebrew/Library/Homebrew/test.rb /usr/local/Homebrew/Library/Taps/homebrew/homebrew-core/Formula/tarsnap.rb --verbose
==> /usr/local/Cellar/tarsnap/1.0.39/bin/tarsnap -c --dry-run /tmp/tarsnap-test-20170821-9566-xa8gv4
dyld: lazy symbol binding failed: Symbol not found: _clock_gettime
  Referenced from: /usr/local/Cellar/tarsnap/1.0.39/bin/tarsnap
  Expected in: /usr/lib/libSystem.B.dylib

dyld: Symbol not found: _clock_gettime
  Referenced from: /usr/local/Cellar/tarsnap/1.0.39/bin/tarsnap
  Expected in: /usr/lib/libSystem.B.dylib
```

The test block runs
```
  test do
    system bin/"tarsnap", "-c", "--dry-run", testpath
  end
```

https://jenkins.brew.sh/job/Homebrew%20Core%20Pull%20Requests/6974/version=el_capitan/console
Duplicated here since that log will eventually be deleted: https://gist.github.com/ilovezfs/8ee666cbb4a5ec36e615771f423db925

I can work around this issue with the following silly hack:
```
    if MacOS.version == :el_capitan && MacOS::Xcode.installed? && MacOS::Xcode.version >= "8.0"
      inreplace "libcperciva/util/monoclock.c", "CLOCK_MONOTONIC",
                                                "UNDEFINED_GIBBERISH"
    end
```

The immediate problem here is that even though there are configure time checks for the availability of clock_gettime, the code in `libcperciva/util/monoclock.c` is guarded with `CLOCK_MONOTONIC` rather than `HAVE_CLOCK_GETTIME`:
```
#ifdef CLOCK_MONOTONIC
        if (clock_gettime(CLOCK_MONOTONIC, &tp) == 0) {
                tv->tv_sec = tp.tv_sec;
                tv->tv_usec = tp.tv_nsec / 1000;
        } else if ((errno != ENOSYS) && (errno != EINVAL)) {
                warnp("clock_gettime(CLOCK_MONOTONIC)");
                goto err0;
        } else
#endif
```

So even though in the Homebrew backend we try to cheat and tell configure the right answer is `ac_cv_func_clock_gettime=no`, this doesn't have the desired effect because `CLOCK_MONOTONIC` is in fact defined in the Xcode 8.x SDK and tarsnap doesn't check whether `HAVE_CLOCK_GETTIME` before proceeding to use `clock_gettime`.

**Comment by gperciva** [comment]
Interesting!  I spent a bit of time poking around, but no results yet.  I see that tarsnap compiles normally from git (and presumably from a tarball as well), so this is quite puzzling.

Note to self: to reproduce, you need to edit
```
/usr/local/Homebrew/Library/Taps/homebbrew/homebrew-core/Formula/tarsnap.rb
```
to remove the replacement.

**Comment by cperciva** [comment]
I believe clock_gettime is provided by librt; is there a library problem here?

**Comment by ilovezfs** [comment]
clock_gettime(3) is found in libc (/usr/lib/system/libsystem_c.dylib) on macOS 10.12+. There's no seperate librt. It's not available on < 10.12.

To reproduce the issue outside of Homebrew on 10.11, set SDKROOT=/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk and MACOSX_DEPLOYMENT_TARGET=10.11

**Comment by gperciva** [comment]
Right, this looks like a simple POSIX violation:

```
#include <time.h>
int main() {
        struct timespec ts;
        int time = clock_gettime(CLOCK_REALTIME, &ts);
}
```

Attempt:
```
Grahams-Mac-mini:tmp gperciva$ uname -a
Darwin Grahams-Mac-mini.local 15.6.0 Darwin Kernel Version 15.6.0: Sun Jun  4 21:43:07 PDT 2017; root:xnu-3248.70.3~1/RELEASE_X86_64 x86_64
Grahams-Mac-mini:tmp gperciva$ clang --version
Apple LLVM version 8.0.0 (clang-800.0.42.1)
Target: x86_64-apple-darwin15.6.0
Thread model: posix
InstalledDir: /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin
Grahams-Mac-mini:tmp gperciva$ clang gettime.c 
gettime.c:4:13: warning: implicit declaration of function 'clock_gettime' is
      invalid in C99 [-Wimplicit-function-declaration]
        int time = clock_gettime(CLOCK_REALTIME, &ts);
                   ^
gettime.c:4:27: error: use of undeclared identifier 'CLOCK_REALTIME'
        int time = clock_gettime(CLOCK_REALTIME, &ts);
                                 ^
1 warning and 1 error generated.
Grahams-Mac-mini:tmp gperciva$ grep "clock_gettime" /usr/include/* -R
/usr/include/php/main/php_config.h:/* do we have clock_gettime? */
Grahams-Mac-mini:tmp gperciva$ 
```

Even though `clock_realtime()` not an optional part of http://pubs.opengroup.org/onlinepubs/9699919799/basedefs/time.h.html

I'll come up with a fix.

**Comment by ilovezfs** [comment]
>Right, this looks like a simple POSIX violation:

On the bright side >= macOS 10.12 doesn't suffer from the same :face_with_head_bandage:

**Comment by cperciva** [comment]
@ilovezfs Dumb question here, but on OS X 10.11, are we trying to link to librt?  Because if that's where clock_gettime is, we *should* be linking to it...

**Comment by gperciva** [comment]
`clock_gettime` does not exist in `<time.h>` on OS X 10.11:
https://github.com/Tarsnap/tarsnap/issues/286#issuecomment-324072189

Recall that there's two issues happening here:
- native OSX 10.11 and earlier: no `clock_gettime()` at all.  (yes, this is a POSIX violation.  We didn't notice this earlier because it also didn't have `CLOCK_REALTIME`, so we never tried to use `clock_gettime`.  I still think it's worth changing libcperciva and POSIX to detect the lack of `clock_realtime()` directly, instead of relying on detecting the lack of `CLOCK_REALTIME`.)
- OSX 10.11 with current XCode: they ship a `<time.h>` aimed at 10.12, so this *does* include `clock_gettime()`.  But the 10.11 libraries still don't include `clock_realtime`, because that wasn't added then.

**Comment by gperciva** [comment]
This was fixed in #290, but the fix doesn't exist in a tarball release yet.
